packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts

```diff
@@ -0,0 +1,219 @@
+			// Get the root dataStore from the detached container.
+			const defaultDataStore = (await container.getEntryPoint()) as TestFluidObject;
+			return {
+				container,
+				defaultDataStore,
+			};
+		}
+
+		function createTestLoader(): Loader {
+			const factory: TestFluidObjectFactory = new TestFluidObjectFactory([
+				[sharedStringId, SharedString.getFactory()],
+				[sharedMapId, SharedMap.getFactory()],
+			]);
+			const codeLoader = new LocalCodeLoader([[codeDetails, factory]], {});
+			const testLoader = new Loader({
+				urlResolver: provider.urlResolver,
+				documentServiceFactory: wrapObjectAndOverride<IDocumentServiceFactory>(
+					provider.documentServiceFactory,
+					{
+						createContainer: () => assert.fail("fail on attach"),
+					},
+				),
+				codeLoader,
+				logger: provider.logger,
+				configProvider: {
+					getRawConfig: (name) =>
+						name === "Fluid.Container.RetryOnAttachFailure" ? true : undefined,
+				},
+			});
+			loaderContainerTracker.add(testLoader);
+			return testLoader;
+		}
+
+		const createPeerDataStore = async (containerRuntime: IContainerRuntimeBase) => {
+			const dataStore = await containerRuntime.createDataStore(["default"]);
+			const peerDataStore = (await dataStore.entryPoint.get()) as ITestFluidObject;
+			return {
+				peerDataStore,
+				peerDataStoreRuntimeChannel: peerDataStore.channel,
+			};
+		};
+
+		beforeEach("createLoader", async function () {
+			provider = getTestObjectProvider();
+			if (
+				semver.compare(provider.driver.version, "0.46.0") === -1 &&
+				(provider.driver.type === "routerlicious" || provider.driver.type === "tinylicious")
+			) {
+				this.skip();
+			}
+			const documentId = createDocumentId();
+			request = provider.driver.createCreateNewRequest(documentId);
+			loader = createTestLoader();
+		});
+
+		afterEach("resetLoaderContainerTracker", () => {
+			loaderContainerTracker.reset();
+		});
+
+		const tests = () => {
+			it("Can serialize detached container", async () => {
+				const { container } = await createDetachedContainerAndGetEntryPoint();
+
+				const snapshotTree = container.serialize();
+
+				const container2 =
+					await loader.rehydrateDetachedContainerFromSnapshot(snapshotTree);
+
+				// Check for default data store
+				const entryPoint = await container2.getEntryPoint();
+				assert.notStrictEqual(entryPoint, undefined, "Component should exist!");
+				// Might need some other asserts, but not sure what
+			});
+			it("Can serialize detached container with data stores after failed attach", async () => {
+				const { container, defaultDataStore } =
+					await createDetachedContainerAndGetEntryPoint();
+
+				// go into attaching state
+				await container.attach(request).then(
+					() => assert.fail("should fail"),
+					() => {},
+				);
+				assert.equal(
+					container.attachState,
+					AttachState.Attaching,
+					"Container should be attaching",
+				);
```

Comment:
maybe move this into createDetachedContainerAndGetEntryPoint, so all the tests validat an attaching state